### ISSUE TITLE
Evaluate integrating drush deploy:hook or patch in core to allow code execution after configuration is imported

### ISSUE THREAD

**Issue by marcoscano** [Lullabot/drainpipe#32]
Currently there is no core mechanism that allows developers to execute an update hook _after_ configuration has been imported. There was once https://www.drupal.org/project/drupal/issues/2901418 but it was closed in favor of integrating drush in core ( https://www.drupal.org/project/ideas/issues/3159647 ). In practice, all our projects already require drush/drush, so my suggestion here is:

- Declare a hard dependency on `drush/drush:^10.3.0` , which is the [version](https://github.com/drush-ops/drush/releases/tag/10.3.0) where the `drush deploy` and `drush deploy:hook` commands were added.

Having that in place, we can then extend the `update` task to include `drush deploy:hook` at the very end. This allows updates such as programmatically copying data into a field that is being created in the same deployment/update task.

**Comment by deviantintegral** [comment]
The last time I ran into a "create and migrate a field" case, there was enough data that as @m4olivei said in https://www.drupal.org/project/drupal/issues/2901418#comment-13168318 that a queue was a better solution.

So, I'm curious - are there use cases other than create and migrate where this hook is useful?

Technically, I don't see much of a reason to have the drush 10.3+ dependency, and calling the task if it's not used by anything will mean it's a no-op? In other words, even if there's different approaches on the code side, the harm seems minimal in adding this with some clear existing benefit.

So.... @justafish ?

**Comment by marcoscano** [comment]
While using a queue is totally fine (and would actually be preferable if there is an unreasonably large amount of data to be processed), I feel conceptually it doesn't fit very well, because we would be writing a plugin that would be intended to ever be used only once. After it's done its job, that's code that would be lingering in the codebase without much purpose. For one-off operations, update hooks are a well-established pattern in Drupal.

Here's a list of recent real-world examples where we used `drush deploy:hook` on New Relic:

* [Re-save all media entities to have metadata fields (deployed at the same time) automatically populated](https://github.com/Lullabot/newrelic/blob/main/web/modules/custom/nr_core/nr_core.deploy.php#L217)
* [Programmatically create a node that uses config from a module that is being installed in the same deployment](https://github.com/Lullabot/newrelic/blob/main/web/modules/custom/nr_node_blog/nr_node_blog.deploy.php#L66)
* [Change a vocabulary name, re-create terms, and re-assign node field values to the new TIDs](https://github.com/Lullabot/newrelic/blob/main/web/modules/custom/nr_node_blog/nr_node_blog.deploy.php#L213)
* [Update path aliases on a small number of nodes after changing their patterns](https://github.com/Lullabot/newrelic/blob/main/web/modules/custom/nr_node_news/nr_node_news.deploy.php#L62)
* [Programmatically translate taxonomy terms that are created programmatically in the same deployment](https://github.com/Lullabot/newrelic/blob/main/web/modules/custom/nr_node_blog/nr_node_blog.deploy.php#L322)

**Comment by justafish** [comment]
> Currently there is no core mechanism that allows developers to execute an update hook after configuration has been imported

You can do this in a post update hook, which will run after configuration is imported. See https://github.com/Lullabot/drainpipe/blob/main/tasks/drupal.yml#L34 and https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_post_update_NAME/9.2.x

**Comment by marcoscano** [comment]
Oh! I wasn't aware of the option `drush updb --no-post-updates` and the pattern:

```
drush updatedb --no-post-updates
drush config:import
drush updatedb --no-cache-clear
```

Conceptually, it feels like a workaround to me, since we will be executing `updb` twice, but I agree the end result will be the same, and this seems to have been added back in 9.0.0-beta8, so it's fairly safe to use at this point.

Thanks for pointing it out, we can close this as won't fix then.

**Comment by deviantintegral** [comment]
Clarification: `--no-post-updates` was added in Drush 9, not Drupal 9.
